lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+
+        result.append(
+            {
+                "name": ".".join(placeholder[:-1]),
+                "arch": ".".join(placeholder[-1:]),
+                "epoch": epoch,
+                "version": version,
+                "release": "".join(buffer),
+                "repo": pkg_repo[1:] if pkg_repo.startswith("@") else pkg_repo,
+                "state": state[cache_type],
+            }
+        )
+
+    return result
+
+
+class _Cache:
+    """Track the current state of the DNF package cache."""
+
+    def __new__(cls) -> "_Cache":
+        """Create new  _Cache object.
+
+        Returns:
+            (_Cache): New _Cache object.
+        """
+        if not hasattr(cls, f"{cls.__name__}__instance"):
+            cls.__instance = super(_Cache, cls).__new__(cls)
+            cls.__instance.__initialized = False
+        return cls.__instance
+
+    def __init__(self) -> None:
+        if self.__initialized:
+            return
+
+        self.__initialized = True
+        self.refresh()
+
+    @property
+    def installed(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get installed packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of installed packages.
+        """
+        return copy.deepcopy(self.__installed)
+
+    @property
+    def available(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get available packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of available packages.
+        """
+        return copy.deepcopy(self.__available)
+
+    def refresh(self) -> None:
+        """Refresh current package cache."""
+        with ThreadPoolExecutor() as executor:
+            installed = executor.submit(_load_cache, "installed")
+            available = executor.submit(_load_cache, "available")
+            self.__installed = installed.result()
+            self.__available = available.result()
+
+
+def _refresh(func: Callable) -> Callable:
+    """Refresh cache after function call.
+
+    Args:
+        func (Callable): Function to wrap.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> None:
+        func(*args, **kwargs)
+        _Cache().refresh()
+
+    return wrapper
+
+
+def _search(
+    cache: List[Dict[str, Union[str, _PackageState]]], name: str
+) -> Optional[Dict[str, Union[str, _PackageState]]]:
+    """Search DNF package cache.
+
+    Args:
+        cache (List[Dict[str, Union[str, _PackageState]]]):
+            DNF package cache to search.
+        name (str):
+            Name of package to locate.
+
+    Returns:
+        (Optional[Dict[str, Union[str, _PackageState]]]):
+            Located cache entry. Returns None if package is not found in cache.
+    """
+    for entry in cache:
+        if entry["name"] == name:
+            return entry
+
+    return None
+
+
+class _MetaDNF(type):
+    """Metaclass for `dnf`.
```

Comment:
This feels like an over-complication to me, consider if this is an approach that brings you significant value over a more simple approach

Reply:
I used this approach as a forcing function to inform the end-user that the charm base they are using does not support DNF. I used a metaclass so that this check/error being raised would be done at the dnf class definition rather than needing to have a conditional be evaluated every time the end-user invokes a dnf command.

I also used the metaclass because I wanted the dnf class to have a `__getitem__`-like syntax for pulling up information about packages, and I wanted to have a class property for retrieving the version of the dnf binary.

Reply:
Removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb